[Scripting] Initial Type Provider API

nerdsupremacist opened the pull request:
This PR implements our basic initial Type Providers API. 

It includes:

- A Code Generation DSL
- An API to convert the results of the Code Generation DSL into a Artifacts that can be imported to a script
- An API for Annotation Based Type Providers using the DSL

## Code Generation DSL

The entire Code Generation DSL is based on the `GeneratedCode` interface:
```kotlin
interface GeneratedCode {
    fun Builder.body()
}
```

This allows users to build reusable constructions for generating code, without being exposed to any internals.

The rest of our DSL is a set of internal extension functions on `GeneratedCode.Builder` to cover the scenarios we need to cover. These include:
- Data Classes
- Objects
- Constants
- Companion Objects
- Functions
- Methods
- Computed Properties (with getters and optionally setters), also available as extensions
- Overriding members
- Enums
- Interfaces
- Inline code as string (as a fallback)

Example:
```kotlin
val myCode = GeneratedCode {
    dataClass("Foo") {
        property<Int>("bar")
        implement(MyInterface::class)

        method("baz") { argument: Int -> 
            val bar by property<Int>()
            // Do something
        }

        override {
            method("other") { first: Int, second: String -> 
                // Call baz using invoke on string
                val bazResult = "baz"(first)

                // Do something else
            }   
        }
    }
}
```

## Using Generated Code

Generated Code can be turned into artifacts, which will return a list of Scripts that should be imported along with a set if imports that should be added to the scripts

```kotlin
val code = GeneratedCode {
    ...
}

val artifacts = code.artifacts()
```

These artifacts can also we included in a `ScriptCompilationConfiguration`:

```kotlin
val newCompilationConfiguration = config.provide(artifacts)

// or

val newCompilationConfiguration = config.provide {
    // Full access to Code Generation DSL
    constant("meaning", 42)
}
```

## Annotation Based Type Providers

To create a Type Provider you define an annotation and implement `AnnotationBasedTypeProvider`

Example:

```kotlin

@Target(AnnotationTarget.FILE)
@Repeatable
@Retention(AnnotationRetention.SOURCE)
annotation class MyAnnotation(val number: Int)

object MyTypeProvider : AnnotationBasedTypeProvider<MyAnnotation> {
    override fun invoke(
        collectedAnnotations: List<ScriptSourceAnnotation<JSON>>,
        context: TypeProvider.Context
    ): ResultWithDiagnostics<GeneratedCode> {
        val number = collectedAnnotations.first().annotation.number 

        return GeneratedCode {
             dataClass("MyClass") {
                 property<Int>("number", default = number)
             }
        }.asSuccess()
    }
}
```

To use a type provider you add it to your script compilation configuration using `typeProviders`.
Example:

```kotlin
ScriptCompilationConfiguration {
    ...
    typeProviders {
        +MyTypeProvider
    }
}
```

## Future Work

- [ ] Caching of the generated code.
- [ ] The annotation based type providers don't check for basic name clashes before compiling to report on each annotation. These errors would then get caught by the compiler, but the error message might not be very helpful to the user

The code generation still doesn't support these scenarios
- [ ] Include a way to add default arguments to functions and methods
- [ ] Subclassing
- [ ] Sealed classes
- [ ] Generics on functions. We already support the receiver type to have a generic parameter, however this parameter can't be a generic parameter of the function, i.e.: `fun <T> Foo<T>.doSomething(): T`

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+/**
+ * Builder for a provided type whose JVM Name can be set
+ */
+interface JVMNameableBuilder {
```
Why is it JVM-specific?
also, the function seems unused, so the intent is not clear.

nerdsupremacist replied:
This was me preparing for something like Data sciences or the shell type provider's extension function heavy generated code. It will often result in clashing names. So renaming the JVM name for things seems appropriate

nerdsupremacist replied:
Removed. Will add it again as a modifier just like override in case it becomes necessary again

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+/**
+ * Add a companion object to the generated type
+ */
+fun GeneratedTypeBuilder.companionObject(name: String? = null, init: ObjectBuilder.() -> Unit) {
```
It would be nice to have tests that use all functions from the DSL, otherwise, they may rot over time and we may discover it too late.

nerdsupremacist replied:
Added a bunch of tests. It should be just shy of 70% coverage now

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+    /**
+     * Add code to the namespace of the Type
+     */
+    fun inner(code: GeneratedCode)
```
I think this is a confusing name, the word `inner` has a different meaning in the Kotlin code.
In the compiler, we're using `child`/`children`  in such places.

nerdsupremacist replied:
Got rid of this abstraction by making `ClassLikeBuilder` implement `GeneratedCode.Builder`

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(vararg code: GeneratedCode) = composite(code.toList())
```
Probably `compose` would be better.

nerdsupremacist replied:
Renamed to `invoke`

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+    /**
+     * Implement a generated interface
+     */
+    fun implement(generated: GeneratedInterface)
```
`implements` would look better for me

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+/**
+ * Builder for a provided type. It can implement an interface and have some code inside.
+ */
+interface GeneratedTypeBuilder : IdentifiableMember {
```
This seems to represent an analog of `ClassLike` things in the compiler, so maybe it makes sense to name it accordingly. `Type` is a bit more general.

nerdsupremacist replied:
Renamed to ClassLike

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+/**
+ * Builder for a provided data class. It can implement interfaces, have code inside and have a number of properties
+ */
+interface DataClassBuilder : JVMNameableBuilder, GeneratedTypeBuilder {
```
I'm not sure that there are benefits of having extracted interfaces from the leaf nodes. Having the public open builder inside the `DataClass` (in this case) will make it leaner, and also allow to reuse some functionality from the similar builders and containers.

nerdsupremacist replied:
Reusing `ClassLike.Builder`
`DataClassBuilder` is still there. But the implementation reuses as much as possible from the `ClassLike.Builder`

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeMethod.kt:
```diff
@@ -0,0 +1,318 @@
+import kotlin.script.experimental.typeProviders.generatedCode.internal.InternalGeneratedCode
+import kotlin.script.experimental.typeProviders.generatedCode.override
+
+private val storedValues = mutableMapOf<String, Any>()
```
this is an unacceptable global state. You need to put it into some context carried with the script.

nerdsupremacist replied:
Replaced global state with serialized values file

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeInlineClass.kt:
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
```
I guess it should belong to a different commit.

nerdsupremacist replied:
Oh yeah. I'll redo the entire git history after we our call

nerdsupremacist replied:
Redid history

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeLazyProperty.kt:
```diff
@@ -0,0 +1,129 @@
+import kotlin.script.experimental.typeProviders.generatedCode.override
+
+
+private val storedValues = mutableMapOf<String, Any>()
```
Global state.

nerdsupremacist replied:
Removed

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeLazyProperty.kt:
```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
```
wrong commit

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeProperty.kt:
```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
```
wrong commit

nerdsupremacist replied:
Reordered

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeTypeAlias.kt:
```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
```
wrong commit

ligee commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeMethod.kt:
```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
```
The approach here looks strange to me, especially in the lack of the usages to chack the intentions. Let's discuss it online.

ligee reviewed (changes requested):
There are quite many comments already, so I think that a big refactoring is imminent. So, I think it doesn't make sense to dive too deep into details now. Let's discuss the approaches online.

nerdsupremacist commented on libraries/scripting/type-providers/build.gradle.kts:
```diff
@@ -0,0 +1,55 @@
+runtimeJar()
+sourcesJar()
+javadocJar()
+testsJar()
```
My bad. I copied and pasted stuff until it worked.

I'll clean up the build.gradle ASAP.

nerdsupremacist replied:
I think it's a lot cleaner now, but I'd love to go over it with you if you don't mind

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/makeDataClass.kt:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @param name Name of your data class
+ */
+fun makeDataClass(name: String, init: DataClassBuilder.() -> Unit): IdentifiableCode {
```
`makeInterface` is definitely planned. I wasn't sure if `makeRegularClass` is necessary though

nerdsupremacist replied:
Renamed to `dataClass`. Put it in the DSL and included ``interface`` as well

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/impl/generatedCodeFromFile.kt:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Include jar file as generated code.
+ */
+fun File.asGeneratedJar(packageName: String): GeneratedCode =
```
I'll write some tests. But this is in case the type provider has an already compiled jar they want to include.

I'm not sure if it will come up. I'm fine with removing it for now and seeing if it comes up

nerdsupremacist replied:
Removed since it was a bit of an additional case. If at some point we need it, I'll add it back

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+     * @return An instance of [GeneratedCode].
+     * Important: don't return an infinite recursion in the form of return `this`
+     */
+    fun body(): GeneratedCode
```
This is a new design I'm trying out, where the user can define their own `GeneratedCode`, but it has to be a composition of the code we provide.

It's inspired by a new API in Swift for UI programming: https://developer.apple.com/documentation/swiftui/view

So the idea is that versions that are not internal have to at some point be reduced to a collection of internal constructs that we can work with

nerdsupremacist replied:
But you're right. Perhaps it should throw an exception instead of returning `this` in that moment

nerdsupremacist replied:
Changed it to `GeneratedCode.Builder.body()` where user's can call our DSL to include code themselves

nerdsupremacist committed:
[Scripting] [Type Providers] Add type providers module

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add base structure for code generation internals

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add implementation for combining generated code

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add standard generated code builder

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add identifiable member interface to annotate types

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add basic overriding capabilities

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add basic class like structure

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add visitor helpers

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add basic interface builder dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add code generation for constants

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add data class code generation dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add enum code generation dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add basic import functionality to code generation dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to inline strings as code

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to include a prebuilt file to the generated code

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add string builder utils

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add composite code visitors

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add the functionality to turn a serializable value into code via deserialization

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add a way to create identifiable members from types and strings

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Allow to create constants with runtime values

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Allow data class properties to be of a concrete type with a reified parameter

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to implement from interfaces known at runtime (using reified parameters)

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to implement an interface and use override for the members in a single statement

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to run a function upon start to the code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add functionality to add computer properties to the code generation dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support for using concrete types for methods used in the interface code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support for declaring functions in the code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support for adding lazy properties in he code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support for methods to the code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support for more combinations for identifiable members

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add support to provide generated code using the DSL to a script compilation configuration

nerdsupremacist committed:
[Scripting] Adding KClass support to filterByAnnotationType util function

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] [Test] Add test structure for testing code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] [Test] Add tests for code generation DSL

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] [Test] Add more complex code generation scenario as its own test class

nerdsupremacist committed:
[Scripting] [Type Providers] [Test] Add base class for testing type providers

nerdsupremacist committed:
[Scripting] [Type Providers] [Test] Add simple sample test for annotation based type providers

nerdsupremacist commented on libraries/scripting/type-providers/build.gradle.kts:
```diff
@@ -0,0 +1,55 @@
+    testCompile(project(":compiler:backend.js"))
+
+    testImplementation(intellijCoreDep()) { includeJars("intellij-core") }
+    testRuntimeOnly(intellijDep()) { includeJars("jps-model") }
```
You're right

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/TypeProvider.kt:
```diff
@@ -0,0 +1,80 @@
+ *
+ * To include a Type Provider in a Scripts definition see [ScriptCompilationConfiguration.Builder.typeProviders].
+ */
+interface TypeProvider<A : Annotation> {
```
As discussed we renamed it to `AnnotaionBasedTypeProvider` and allow to create the necessary artifacts from the code generation DSL

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/TypeProvider.kt:
```diff
@@ -0,0 +1,80 @@
+ * To include a Type Provider in a Scripts definition see [ScriptCompilationConfiguration.Builder.typeProviders].
+ */
+interface TypeProvider<A : Annotation> {
+    data class Context internal constructor(
```
Done

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/TypeProvider.kt:
```diff
@@ -0,0 +1,80 @@
+     * @return In case of success the list of annotations that should be turned into types.
+     * In case of a failure an appropriate message for the developer.
+     */
+    fun preProcess(
```
Done

nerdsupremacist commented on libraries/scripting/type-providers/src/kotlin/script/experimental/typeProviders/generatedCode/GeneratedCode.kt:
```diff
@@ -0,0 +1,302 @@
+        /**
+         * Create a composition of multiple instances of Generated Code
+         */
+        fun composite(iterable: Collection<GeneratedCode>): GeneratedCode = if (iterable.count() == 1)
```
Went with invoke

nerdsupremacist committed:
[Scripting] [Type Providers] [JSON] Add Example JSON type providers

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add object code generation dsl

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Add mechanism to convert GeneratedCode to a set of artifacts

With this commit we are now able to convert instances Generated Code into files and imports that should be applied to a script

This also provides our standard implementation of Generated Code Visitor

nerdsupremacist committed:
[Scripting] [Type Providers] [Code Generation DSL] Adding basic public generated code interface

nerdsupremacist committed:
[Scripting] [Type Providers] Add API to create type providers based on annotation usage
